### PR TITLE
Issue #273: Resolved the streaming write issue for spark 3.x.

### PR DESCRIPTION
**Pull request by himanshukohli09**
This issue was beacuse of breaking changes of ExpressionEncoder from spark 2.x to 3.x.

### CODE REVIEW

File: connector/src/main/scala/com/google/cloud/spark/bigquery/BigQueryStreamWriter.scala [hunk 1]
```diff
@@ -39,17 +36,13 @@ private[bigquery] object BigQueryStreamWriter extends Logging {
...
-      )
-    )
-    // Create fixed dataframe
-    val dataFrame: DataFrame = sqlContext.createDataFrame(rowRdd, schema)
-    val table = Option(bigQueryClient.getTable(opts.getTableId))
 
+    // using createDataFrame(List[Row], schema) instead of
+    // createDataFrame(RDD[Row], schema) as for creating RDD[Row] we need to use
+    // ExpressionEncoder, which has breaking changes from spark 2.x to 3.x
+    val dataFrame = sqlContext.createDataFrame(data.collectAsList(), data.schema)
```
Comment (davidrabinowitz) [inline connector/src/main/scala/com/google/cloud/spark/bigquery/BigQueryStreamWriter.scala:43]: `collectAsList()` collects all the data to the driver, and as such it cannot be used for large DataFrames. We need another approach that work on the data in the partitions

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: connector/src/main/scala/com/google/cloud/spark/bigquery/BigQueryStreamWriter.scala (hunk 1 above)
Comment (davidrabinowitz) [reply connector/src/main/scala/com/google/cloud/spark/bigquery/BigQueryStreamWriter.scala:43]: Based on https://github.com/apache/spark/commit/fab4ca5156d5e1cc0e976c7c27b28a12fa61eb6d I'd try the `ExpressionEncoder.createDeserializer()`. Notice you can have different approach per spark version.

File: spark-3/src/main/java/com/google/cloud/spark/bigquery/spark3/DataFrameToRDDConverter.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.google.cloud.spark.bigquery.spark3;
```
Comment (andr0267) [inline spark-3/src/main/java/com/google/cloud/spark/bigquery/spark3/DataFrameToRDDConverter.java:1]: Minor and maybe not relevant: maybe it makes sense to move that interface under bigquery.connector.common - as it defines behaviour of not only spark3.

File: build.sbt [hunk 3]
```diff
@@ -62,9 +62,20 @@ lazy val commonTestDependencies = Seq(
   "com.google.truth" % "truth" % "1.0.1" % "test"
 )
 
+lazy val spark3 = (project in file("spark-3"))
```
Comment (davidrabinowitz) [inline build.sbt:65]: Please rename to `spark3support`

File: build.sbt [hunk 4]
```diff
@@ -62,9 +62,20 @@ lazy val commonTestDependencies = Seq(
...
+  .enablePlugins(BuildInfoPlugin)
+  .settings(
+    commonSettings,
+    name := "spark-3",
```
Comment (davidrabinowitz) [inline build.sbt:69]: Please rename to `spark3support`

File: spark-3/src/main/java/com/google/cloud/spark/bigquery/spark3/DataFrameToRDDConverter.java (hunk 2 above)
Comment (davidrabinowitz) [reply spark-3/src/main/java/com/google/cloud/spark/bigquery/spark3/DataFrameToRDDConverter.java:1]: Thanks for the comment - `bigquery.connector.common` is spark agnostic, hence this code cannot reside there.

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
Can you please add an acceptance test, or share test result from spark 3?

File: build.sbt (hunk 4 above)
Comment (himanshukohli09) [reply build.sbt]: done

File: build.sbt (hunk 3 above)
Comment (himanshukohli09) [reply build.sbt]: done

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
Acceptance test failed on 
```
[error] Test com.google.spark.bigquery.acceptance.DataprocImage20AcceptanceTest.testBigNumeric failed: expected: DONE
[error] but was : ERROR, took 29.064 sec
```

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
I believe there's a better way to solve the python lib issue - all the GCP python library use the `google.cloud` namespace, so there's no reason we won't be able to use `google.cloud.spark.bigquery`. Please take a look at the [BigQuery](https://github.com/googleapis/python-bigquery), [PubSub](https://github.com/googleapis/python-pubsub) and [others](https://github.com/googleapis?q=python-&type=&language=python&sort=). Also, please create a separate PR for this, it should not be part of the Spark 3 support.

**Comment by himanshukohli09** [comment]
> I believe there's a better way to solve the python lib issue - all the GCP python library use the `google.cloud` namespace, so there's no reason we won't be able to use `google.cloud.spark.bigquery`. Please take a look at the [BigQuery](https://github.com/googleapis/python-bigquery), [PubSub](https://github.com/googleapis/python-pubsub) and [others](https://github.com/googleapis?q=python-&type=&language=python&sort=). Also, please create a separate PR for this, it should not be part of the Spark 3 support.

Done, this has been resolved with https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/428

**Comment by himanshukohli09** [comment]
> Acceptance test failed on
> 
> ```
> [error] Test com.google.spark.bigquery.acceptance.DataprocImage20AcceptanceTest.testBigNumeric failed: expected: DONE
> [error] but was : ERROR, took 29.064 sec
> ```

This is got fixed after rebasing this branch with master (containing the BigNumeric acceptance test fix: PR https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/428)

**Review by davidrabinowitz — APPROVED** [review]

**Comment by milicevica23** [comment]
Hi i have similar error, as i see the code uses com.google.cloud.spark.bigquery.spark3 but in some utils the error is trown? 
`
java.lang.NoSuchMethodError: 'org.apache.spark.sql.catalyst.encoders.ExpressionEncoder org.apache.spark.sql.catalyst.encoders.RowEncoder.apply(org.apache.spark.sql.types.StructType)'
        at org.apache.spark.sql.PreScala213SparkSqlUtils.createExpressionEncoder(PreScala213SparkSqlUtils.java:53)
        at com.google.cloud.spark.bigquery.spark3.Spark3DataFrameToRDDConverter.convertToRDD(Spark3DataFrameToRDDConverter.java:40)
        at com.google.cloud.spark.bigquery.BigQueryStreamWriter$.writeBatch(BigQueryStreamWriter.scala:50)
        at com.google.cloud.spark.bigquery.BigQueryStreamingSink.addBatch(BigQueryStreamingSink.scala:53)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runBatch$17(MicroBatchExecution.scala:732)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
        at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
        at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runBatch$16(MicroBatchExecution.scala:729)
        at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:427)
        at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:425)
        at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runBatch(MicroBatchExecution.scala:729)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$2(MicroBatchExecution.scala:286)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:427)
        at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:425)
        at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$1(MicroBatchExecution.scala:249)
        at org.apache.spark.sql.execution.streaming.ProcessingTimeExecutor.execute(TriggerExecutor.scala:67)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runActivatedStream(MicroBatchExecution.scala:239)
        at org.apache.spark.sql.execution.streaming.StreamExecution.$anonfun$runStream$1(StreamExecution.scala:311)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
        at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:289)
        at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.$anonfun$run$1(StreamExecution.scala:211)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
        at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:211)
`

**Pull request merged**